ml4gw/transforms/decimator.py

```diff
@@ -0,0 +1,195 @@
+        schedule: torch.Tensor = None,
+        device=None,
+    ):
+        """
+        Build a mask of indices to decimate a strain according to the schedule.
+        """
+        sample_rate = (
+            sample_rate if sample_rate is not None else self.sample_rate
+        )
+        schedule = schedule if schedule is not None else self.schedule
+        device = device if device is not None else self.device
+
+        idx = torch.tensor([], dtype=torch.long, device=device)
+
+        for s in schedule:
+            if idx.size(0) == 0:
+                start = int(s[0] * sample_rate)
+            else:
+                start = int(idx[-1]) + int(idx[-1] - idx[-2])
+            stop = int(start + (s[1] - s[0]) * sample_rate)
+            step = int(sample_rate // s[2])
+            new_idx = torch.arange(
+                start, stop, step, dtype=torch.long, device=device
+            )
+            idx = torch.cat((idx, new_idx))
+        return idx
+
+    def split_by_schedule(
+        self, strain: torch.Tensor, schedule: torch.Tensor = None
+    ):
+        """
+        Split a strain into segments according to schedule.
+        """
+        schedule = schedule if schedule is not None else self.schedule
+        split_sizes = (
+            ((schedule[:, 1] - schedule[:, 0]) * schedule[:, 2])
+            .long()
+            .tolist()
+        )
+        segments = torch.split(strain, split_sizes, dim=-1)
+        return segments
+
+    def forward(
+        self,
+        strain: torch.Tensor,
+        sample_rate: int = None,
+        schedule: torch.Tensor = None,
+        split: bool = False,
+        device=None,
+    ):
+        """
+        Run decimation: get decimated waveform and/or return split segments.
+
+        Parameters:
+            strain : torch.Tensor
+                Input tensor of shape (B, C, T), where the last dimension
+                is time.
+            sample_rate : int, optional
+                Original sample rate of the strain in Hz. Defaults to
+                self.sample_rate.
+            schedule : torch.Tensor, optional
+                Tensor of shape (N, 3), each row [start, end, target_sr].
+            split : bool, optional
+                If True, returns list of segments split by schedule along
+                the last dim.
+            device : torch.device, optional
+                Device on which to perform indexing.
+
+        Returns:
+            torch.Tensor or List[torch.Tensor]
+                Decimated strain or list of segments if split=True.
+        """
+
+        device = device if device is not None else self.device
+        sample_rate = (
+            sample_rate if sample_rate is not None else self.sample_rate
+        )
+        schedule = schedule if schedule is not None else self.schedule
+
+        self._validate_inputs(sample_rate, schedule, strain=strain)
+
+        if (
+            self.idx is not None
+            and sample_rate == self.sample_rate
```

Comment:
This won't work if a user passes a sample rate initially, but not during the forward call. I would suggest having the pattern be that the sample rate and schedule _must_ be passed on initialization, and then can't be passed during `forward`. If a user wants to do a different decimation, they should create a new `Decimator` object.